controllers/operatorscaler_controller.go

```diff
@@ -0,0 +1,364 @@
+		return err
+	}
+
+	return nil
+}
+
+func (r *OperatorScalerReconciler) scaleOperators() error {
+
+	var returnErr error
+
+	for kind, pkgNames := range kindToPkgNamesMapping {
+
+		crList := &metav1.PartialObjectMetadataList{}
+		crList.TypeMeta = kind.TypeMeta
+
+		err := r.Client.List(r.ctx, crList)
+		if err != nil {
+			if meta.IsNoMatchError(err) {
+				continue
+			}
+			msg := fmt.Sprintf("failed to list %s", kind.TypeMeta.Kind)
+			r.log.Error(err, msg)
+			multierr.AppendInto(&returnErr, err)
+			continue
+		}
+
+		if len(crList.Items) == 0 {
+			continue
+		}
+
+		for _, pkgName := range pkgNames {
+			key := fmt.Sprintf(csvLabelKey, pkgName, r.OperatorNamespace)
+
+			csvList := &oprv1a1.ClusterServiceVersionList{}
+			err = r.Client.List(
+				r.ctx, csvList,
+				client.InNamespace(r.OperatorNamespace),
+				client.MatchingLabels(map[string]string{key: ""}),
+			)
+			if err != nil {
+				r.log.Error(err, "failed to list csvs of label", "label", key)
+				multierr.AppendInto(&returnErr, err)
+				continue
+			}
+
+			err = r.updateCsvsReplicas(csvList.Items...)
+			if err != nil {
+				r.log.Error(err, "failed to update csvs replica")
+				multierr.AppendInto(&returnErr, err)
+			}
+		}
+	}
+
+	return returnErr
+}
+
+func (r *OperatorScalerReconciler) updateCsvsReplicas(csvList ...oprv1a1.ClusterServiceVersion) error {
+
+	var returnErr error
+
+	for idx := range csvList {
+		csv := &csvList[idx]
+		var updateRequired bool
+		for i := range csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs {
+			if *csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs[i].Spec.Replicas != int32(1) {
+				updateRequired = true
+			}
+			csv.Spec.InstallStrategy.StrategySpec.DeploymentSpecs[i].Spec.Replicas = ptr.To(int32(1))
+		}
+
+		if !updateRequired {
+			continue
+		}
+
+		err := r.Client.Update(r.ctx, csv)
+		if err != nil {
+			r.log.Error(err, "failed to update csv replica", "csvName", csv.Name)
+			multierr.AppendInto(&returnErr, err)
+			continue
+		}
+		r.log.Info("csv updated successfully", "csvName", csv.Name)
+	}
+
+	return returnErr
+}
+
+func (r *OperatorScalerReconciler) addWatches(crdName string) error {
+
+	if !slices.Contains(crdNameList, crdName) {
+		return nil
+	}
+
+	if r.watching[crdName] {
+		return nil
+	}
+
+	r.log.Info("adding dynamic watch", "kind", crdName)
+
+	kind := crdNameToKindMapping[crdName]
+	err := r.addDynamicWatch(kind)
+	if err != nil {
+		r.log.Error(err, "failed to add dynamic watch", "kind", crdName)
+		return err
+	}
+
+	r.watching[crdName] = true
+
+	return nil
+}
+
+func (r *OperatorScalerReconciler) addDynamicWatch(kind client.Object) error {
+	return r.controller.Watch(
+		source.Kind(
+			r.cache,
+			kind,
+			&handler.EnqueueRequestForObject{},
+			// Trigger the reconcile for creation events of the object.
+			// This ensures the replicas in the CSV are scaled up based on the presence of Custom Resource (CR).
+			createPredicate,
+			predicate.GenerationChangedPredicate{},
+		),
+	)
```

Comment:
This is only used once. Do we really need a function for this?

Reply:
Removed the func as discussed